megamek/src/megamek/client/RandomNameGenerator.java

```diff
@@ -375,48 +247,191 @@ public int getPercentFemale() {
+        //region Variable Instantiation
+        int numEthnicCodes = 0;
+        //endregion Variable Instantiation
+
+        //region Map Instantiation
+        maleGivenNames = new HashMap<>();
+        femaleGivenNames = new HashMap<>();
+        surnames = new HashMap<>();
+        factionGivenNames = new HashMap<>();
+        factionEthnicCodes = new HashMap<>();
+
+        // Determine the number of ethnic codes
+        File masterAncestryFile = new MegaMekFile(Configuration.namesDir(), FILENAME_MASTER_ANCESTRY).getFile();
+        try (InputStream is = new FileInputStream(masterAncestryFile);
+             Scanner input = new Scanner(is, "UTF-8")) {
+
+            while (input.hasNextLine()) {
+                input.nextLine();
+                numEthnicCodes++;
+            }
+        } catch (IOException e) {
+            System.err.println("RandomNameGenerator.populateNames(): Could not find '" + masterAncestryFile + "'");
         }
-        return rng;
+
+        // Then immediately instantiate the number of weighted maps needed for Given Names and Surnames
+        for (int i = 1; i <= numEthnicCodes; i++) {
+            maleGivenNames.put(i, new WeightedMap<>());
+            femaleGivenNames.put(i, new WeightedMap<>());
+            surnames.put(i, new WeightedMap<>());
+        }
+        //endregion Map Instantiation
+
+        //region Read Names
+        readNamesFileToMap(maleGivenNames, FILENAME_FIRSTNAMES_MALE);
+        readNamesFileToMap(femaleGivenNames, FILENAME_FIRSTNAMES_FEMALE);
+        readNamesFileToMap(surnames, FILENAME_SURNAMES);
+        //endregion Read Names
+
+        //region Faction Files
+        // all faction files should be in the faction directory
+        File factionsDir = new MegaMekFile(Configuration.namesDir(), DIR_NAME_FACTIONS).getFile();
+        String[] fileNames = factionsDir.list();
+
+        if ((fileNames == null) || (fileNames.length == 0)) {
+            //region No Factions Specified
+            System.err.println("RandomNameGenerator.populateNames(): No faction files found!");
+            // We will create a general list where everything is weighted at one to allow player to
+            // play with named characters, indexing it at 1
+            String key = "General";
+
+            // Initialize Maps
+            factionGivenNames.put(key, new HashMap<>());
+            factionEthnicCodes.put(key, new WeightedMap<>());
+
+            // Add information to maps
+            for (int i = 0; i <= numEthnicCodes; i++) {
+                factionGivenNames.get(key).put(i, new WeightedMap<>());
+                factionGivenNames.get(key).get(i).add(1, i);
+                factionEthnicCodes.get(key).add(1, i);
+            }
+            //endregion No Factions Specified
+        } else {
+            for (String filename : fileNames) {
+                // Determine the key based on the file name
+                String key = filename.split("\\.txt")[0];
```

Comment:
This puts a reliance on the filenames ending in .txt (all lower), so I'd add that comment above with the file names.

Reply:
I've added it to the file system documentation, but must have missed it at the top of this file. Fixed.